tests/system/src/test/java/test/com/sun/marlin/ScaleX0Test.java

```diff
@@ -0,0 +1,213 @@
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package test.com.sun.marlin;
+
+import java.io.ByteArrayOutputStream;
+import java.io.PrintStream;
+import java.lang.reflect.Method;
+
+import java.util.Locale;
+import java.util.concurrent.CountDownLatch;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import javafx.application.Application;
+import javafx.application.Platform;
+import javafx.beans.value.ObservableValue;
+import javafx.scene.Group;
+import javafx.scene.Node;
+import javafx.scene.Scene;
+import javafx.scene.control.Button;
+import javafx.scene.control.Slider;
+import javafx.scene.layout.GridPane;
+import javafx.scene.layout.Pane;
+import javafx.scene.paint.Color;
+import javafx.scene.shape.Circle;
+import javafx.scene.shape.Rectangle;
+import javafx.stage.Stage;
+
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import test.util.Util;
+
+/**
+ * @test
+ * @bug 8312603
+ * @summary Check the crash with MarlinFX renderer if scaleX or scaleY is pure 0.0
+ */
+public class ScaleX0Test {
+
+    private final static int SIZE = 800;
+
+    static final ByteArrayOutputStream out = new ByteArrayOutputStream(2048 * 1024);
+    static final PrintStream defaultErrorStream = System.err;
+
+    static {
+        Locale.setDefault(Locale.US);
+
+        System.setProperty("prism.verbose", "false");
+        // Enable Marlin logging
+        System.setProperty("prism.marlin.log", "true");
+    }
+
+    @BeforeClass
+    public static void setupOnce() throws Exception {
+        CountDownLatch startupLatch = new CountDownLatch(1);
+        Util.startup(startupLatch, () -> {
+            Platform.setImplicitExit(false);
+            startupLatch.countDown();
+        });
+    }
+
+    @AfterClass
+    public static void teardownOnce() {
+        Util.shutdown();
+    }
+
+    @Test(timeout = 15000)
+    public void testMarlinAIOOBEwhenScaleXIs0() {
+
+        Scene scene = createScene();
+
+        // Capture stderr:
+        System.setErr(new PrintStream(out, true));
+
+        Platform.runLater(() -> {
+            Stage stage = new Stage();
+            stage.setScene(scene);
+            stage.show();
+        });
+
+        try {
+            Thread.sleep(2000L);
+        } catch (InterruptedException ie) {
+            Logger.getLogger(ScaleX0Test.class.getName()).log(Level.SEVERE, "interrupted", ie);
+        }
+
+        // Restore stderr:
+        System.setErr(defaultErrorStream);
```

Comment:
I missed one more thing earlier.
The sleep of 2 seconds is an uncertain time, which can be made little more specific.
- Add an onShown latch for Stage
- Capture the stderr after stage is shown and keep the sleep of 2 seconds as is

This makes it little clearer that the issue occurs after stage is shown.
Here is the change to replace above code:

```
CountDownLatch showLatch = new CountDownLatch(1);
Platform.runLater(() -> {
    Stage stage = new Stage();
    stage.setScene(scene);
    stage.setOnShown(s -> showLatch.countDown());
    stage.show();
});
Util.waitForLatch(showLatch, 5, "Failed to show the stage.");

// Capture stderr:
System.setErr(new PrintStream(out, true));

try {
    Thread.sleep(2000L);
} catch (InterruptedException ie) {
    Logger.getLogger(ScaleX0Test.class.getName()).log(Level.SEVERE, "interrupted", ie);
}

// Restore stderr:
System.setErr(defaultErrorStream);
```

Please check if this is a correct understanding.
I tested it several times with and without the fix: Always Failed without fix and always passed with fix.